enhancements/authentication/custom-scc-preemption-prevention.md

```diff
@@ -0,0 +1,252 @@
+
+If the hinted SCC does not exist in the system, it is being ignored silently.
+In this case SCC admission continues with all remaining SCCs available in the system as described above.
+
+If the hinted SCC does exist in the system it is being prepended to the sorted list describe above,
+before the highest priority SCC.
+
+This ensures that the hinted SCC is deterministically considered to be applied first
+before any other SCC available in the system.
+
+Pros:
+- Less hard failure modes than the proposed "hard protection" mechanism.
+- Less intrusive than the "hard protection" alternative
+as the workload will continue to be asserted against the remaining SCCs available in the system.
+
+Cons:
+- If the hinted SCC is not applicable to the workload,
+admission continues with asserting all remaining SCCs available in the system as described above.
+- If the hinted SCC is not applicable, admission continues with asserting all the other SCCs available in the system,
+potentially even custom SCCs which still can cause problems.
+- All core OpenShift workloads need to set that annotation in their deployment artifacts.
+This implies that, after the change merges, many repositories have to be touched.
+- To verify that hinted SCCs are really being applied, an e2e test has to be written.
+This can be done centrally.
+
+An existing implementation is available at https://github.com/openshift/apiserver-library-go/pull/105.
+
+This proposal favors the "hard protection" variant and continues to describe that one further on.
+
+### Workflow Description
+
+Opting into this feature for core OpenShift workloads is straight-forward,
+with just an annotation needed to be set in the deployment manifest.
+
+### Variation (optional)
+
+Another variation for both proposed suggestions is to accept a comma-separated list of SCCs.
+This could be helpful in case additional SCCs are going to be introduced and the deployment manifest
+needs to be compatible with both old and new applicable SCCs.
+
+However, as adding a list further adds complexity, this suggestion is being omitted as an variation only
+with the potential of being introduced as a backwards-compatible change in the future.
+
+### API Extensions
+
+A new annotation `openshift.io/required-scc` is introduced.
+
+### Implementation Details/Notes/Constraints [optional]
+
+N/A
+
+### Risks and Mitigations
+
+1. OpenShift workloads that are not applicable to the requested SCC might fail to be admitted.
+Mitigation: The existing e2e test suite in OpenShift will give very early signals of failed admission attempts
+as the workload will be unschedulable.
+
+### Drawbacks
+
+This proposal adds yet another additional complexity on top of the existing SCC logic
+which also already interleaves with pod security admission.
+The mitigation here is to reduce complexity by favoring the "hard protection" variant.
+
+## Design Details
+
+### Open Questions [optional]
```

Comment:
\<warning\>I get a bit philosophical here.\</warning\>

I keep asking myself - why are we doing this? Bear with me:

I would assume that a workload that we control should be capable of setting its security context to the last detail so that it can never be thrown off by any unexpected SCC mismatch. Yet that does not seem to be the case.

What is it that is making us pin SCCs rather than audit our workloads security context's and have them always behave deterministically? In the end, isn't it only generating the SELinux/UID/supplemental groups? Would you know?

It's an interesting thought excercise that I've been struggling with around this topic.

Reply:
good thought 👍 pinning a required SCC is sort of a "shortcut". We are indeed able to audit the workloads in question and craft deployment manifests that will pick the correct SCC. I see the following reasons though why pinning is still useful:

1. generally, SCCs are very confusing, the algorithm quite complex and getting the workload manifests takes time, and potentially an e2e test to get them right. There's a reason why upstream removed equivalent pod security policies.
2. we still can be preempted by a custom SCC if it has the same priority, is equally restrictive but lexically sorted before the intended out-of-the-box SCC. While this won't do a harm it is still confusing to see OpenShift core workloads picking custom SCCs.

2. is cosmetic however 1. (I believe) is valid. It requires discipline and potentially e2e tests to get this right.

Reply:
I assume the important bits are the mutating one, and possibly not all of them. I would like to find specific answers to my questions in the future so that we can possibly derive these bits from the others and separate these two in seek of an easier maintenance.

Reply:
+1 it is this mixture of "restriction" and "mutation" that makes SCCs hard to reason about. There is no clear definition of "more restrictive" in the SCC docs, and in particular it is possible for a "less restrictive" SCC to break an existing workload. e.g. a workload with runAsNonRoot=true and no explicit runAsUser will work under "RunAsRange" (because a user is automatically assigned from the range) but will fail under "RunAsAny" (because that does not assign any user.)  Intuitively you would expect that "RunAsAny" is the _least restrictive_ policy, but since it doesn't offer a default user it is also less "runnable". There needs to be a clear definition of what it means for a pod to "match an SCC" vs. for a pod to be "runnable with an SCC". Currently these two don't align, which is the core of the problem.

Reply:
The user behavior is known and is indeed quite the pain point. We need to identify any behavior that's prone to breaking, and I assume that in the end it would be all of the mutation.

This need for a proper behavior mapping, and possibly proposing a way to deal with it, should not block this enhancement from merging though.

Reply:
>  one reason "just fully specify your security context such that you can only match an appropriate SCC" doesn't work is because you cannot specify a RunAsUser if you want to get RunAsRange behavior(where your pod gets mutated to specify a valid uid within the NS assigned range).

very good point 👍